pkg/vpcmodel/connectivityAbstraction.go

```diff
@@ -22,19 +24,33 @@ import (
 //      nodeSetToNodeSet: connections of   <node     in the nodeSet>  ->  <node     in the nodeSet>
 //      otherFromNodeSet: connections of   <node     in the nodeSet>  ->  <node not in the nodeSet>
 //      otherToNodeSet:   connections of   <node not in the nodeSet>  ->  <node     in the nodeSet>
-// 2. for the last three groups, we verify that the abstraction assumption holds
-//     todo: this check is not complete in code yet, and currently we ignore its result
-// 3. we do the abstraction (for this PR, even if the abstraction assumption does not hold):
+// 2. we do the abstraction on the last three groups:
 //    the connectivity of n->NS is union of all of n->n1, n->n2, n->n3
-//    todo: what to do if the abstraction assumption does not hold?
-
-//nolint:gocritic // for now, we comment out the check calls
-func nodeSetConnectivityAbstraction(nodesConn GeneralConnectivityMap, nodeSet NodeSet) GeneralConnectivityMap {
-	otherToOther, nodeSetToNodeSet, otherFromNodeSet, otherToNodeSet := partitionConnectivityByNodeSet(nodesConn, nodeSet)
-	// checkConnectivityAbstractionValidity(otherFromNodeSet, nodeSet, false)
-	// checkConnectivityAbstractionValidity(otherToNodeSet, nodeSet, true)
-	// checkConnectivityAbstractionValidity(nodeSetToNodeSet, nodeSet, true)
-	return mergeConnectivityWithNodeSetAbstraction(otherToOther, nodeSetToNodeSet, otherFromNodeSet, otherToNodeSet, nodeSet)
+// 3. for the last three groups, we collect the missing connections.
+// 4. we add the 4th group to connectivity
+
+// NodeSetAbstraction abstract nodesets, one after the other
+type NodeSetAbstraction struct {
+	// abstractedConnectivity holds the abstracted connectivity that reflated after the last nodeSet abstraction
+	abstractedConnectivity GeneralConnectivityMap
+}
+
+func newNodeSetAbstraction(nodesConn GeneralConnectivityMap) *NodeSetAbstraction {
+	return &NodeSetAbstraction{nodesConn.copy()}
+}
+
+func (nsa *NodeSetAbstraction) abstractNodeSet(nodeSet NodeSet) *AbstractionInfo {
```

Comment:
NodeSetAbstraction does not know LB

Reply:
I would add a comment in the beginning of this file, that this abstraction is currently only used for LB, though it supports general NodeSet object.

Reply:
I would not
this file should not know about load balancer, not in code, and not in comments
it does not make sense that  if someone would like to use this struct, he should fix its documentation

Reply:
the point is that we do not yet have another use case for using this abstraction other than LB, so it is currently a bit confusing.

Reply:
I agree with Adi; I also had the same comment (also in other places)
Perhaps you can add that at the moment the only relevant abstraction is LB

Reply:
I strongly disagree with both of you.
and you are doing this mistake in a lot of places.
for example, I see a lot of comments in vpcmodel that is ibm specific (and sometime in the code)
these comments will be incorrect when using vpcmodel for other vendors
you will need to wildly update them

encapsulation is an important principle, please try to keep it as much as possible

Reply:
I don't think it breaks encapsulation if you add a comment that this abstraction is currently only relevant to LB objects, which is another abstract type that extends NodeSets.
It is an informative comment so that one can understand the context and motivation for this abstraction.